src/main/java/com/google/devtools/build/lib/hash/Blake3Hasher.java

```diff
@@ -0,0 +1,115 @@
+package com.google.devtools.build.lib.hash;
+
+import com.google.common.hash.HashCode;
+import com.google.common.hash.Hasher;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class Blake3Hasher extends AbstractHasher {
+  // These constants match the native definitions in:
+  // https://github.com/BLAKE3-team/BLAKE3/blob/master/c/blake3.h
+  public static final int KEY_LEN = 32;
+  public static final int OUT_LEN = 32;
+
+  // To reduce the number of calls made via JNI, buffer up to this many bytes.
+  // If a call to "hash()" is made and less than this much data has been
+  // written, a single JNI call will be made that initializes, hashes, and
+  // cleans up the hasher, rather than making separate calls for each operation.
+  public static final int ONESHOT_THRESHOLD = 8 * 1024;
+
+  private final ReentrantLock rl = new ReentrantLock();
+  private ByteBuffer buffer;
+  private long hasher = -1;
+
+  public boolean isAllocated() {
+    return (hasher != -1);
+  }
+
+  private void initOnce() {
+    if (!isAllocated()) {
+      hasher = Blake3JNI.allocate_and_initialize_hasher();
+    }
+  }
+
+  public Blake3Hasher() {}
+
+  private void resetBuffer(int minLength) {
```

Comment:
i remember some pathological cases where you were saying update(one byte), update(one byte), update(one byte)--do you think it makes sense to set a min... min length? for this buffer?  I guess you can't make that size too big, so maybe this is a bad idea, just realizing that you'd be flushing through JNI on every single one of those calls.  really depends on whether or not we think that happens much.